File: packages/companion/swift/Sources/BridgethingCompanion/NowPlayingHub.swift
import BridgethingGateway
import BridgethingGlue
import BridgethingSchema
import Foundation

public final class NowPlayingHub: NowPlayingSink, @unchecked Sendable {
    private final class SourceState {
        var snapshot: PlayerState?
        var appBundle: String = ""
        var hasItem: Bool = false
        var wantsVolume: Bool = false
        var queue: QueueSnapshot?
        var targets: PlaybackTargets?
        var seq: Int64 = 0
    }

    private enum Op: Sendable {
        case player(sourceId: String, snapshot: PlayerState, appBundle: String, hasItem: Bool, wantsVolume: Bool)
        case queue(sourceId: String, queue: QueueSnapshot)
        case targets(sourceId: String, targets: PlaybackTargets)
        case clear(sourceId: String)
        case reconnect
    }

    private let gateway: BridgethingGateway
    private let ops: AsyncStream<Op>
    private let submit: AsyncStream<Op>.Continuation

    private var sources: [String: SourceState] = [:]
    private var seqCounter: Int64 = 0
    private var heldScopes: Set<CompanionAuthorityScope> = []
    private var claimedBundle: String?

    private let lock = NSLock()
    private var consumer: Task<Void, Never>?
    private var currentId: String?
    private var transports: [String: any NowPlayingTransport] = [:]

    public init(gateway: BridgethingGateway) {
        self.gateway = gateway
        var cont: AsyncStream<Op>.Continuation!
        ops = AsyncStream(bufferingPolicy: .unbounded) { cont = $0 }
        submit = cont
    }

    public func start() {
        lock.lock()
        guard consumer == nil else { lock.unlock(); return }
        let stream = ops
        let task = Task { [weak self] in
            for await op in stream {
                guard let self else { return }
                await handle(op)
            }
        }
        consumer = task
        lock.unlock()
    }

    public func stop() {
        lock.lock()
        let task = consumer
        consumer = nil
        lock.unlock()
        task?.cancel()
    }

    public func submitPlayer(
        sourceId: String,
        snapshot: PlayerState,
        appBundle: String,
        hasItem: Bool,
        wantsVolume: Bool
    ) {
        submit.yield(.player(
            sourceId: sourceId,
            snapshot: snapshot,
            appBundle: appBundle,
            hasItem: hasItem,
            wantsVolume: wantsVolume
        ))
    }

    public func submitQueue(sourceId: String, queue: QueueSnapshot) {
        submit.yield(.queue(sourceId: sourceId, queue: queue))
    }

    public func submitTargets(sourceId: String, targets: PlaybackTargets) {
        submit.yield(.targets(sourceId: sourceId, targets: targets))
    }

    public func clearSource(sourceId: String) {
        submit.yield(.clear(sourceId: sourceId))
    }

    public func onConnect() {
        submit.yield(.reconnect)
    }

    public func register(sourceId: String, transport: any NowPlayingTransport) {
        lock.lock()
        transports[sourceId] = transport
        lock.unlock()
    }

    public func unregister(sourceId: String) {
        lock.lock()
        transports.removeValue(forKey: sourceId)
        lock.unlock()
    }

    public func currentTransport() -> (any NowPlayingTransport)? {
        lock.lock()
        defer { lock.unlock() }
        guard let currentId else { return nil }
        return transports[currentId]
    }

    public func currentSource() -> String? {
        lock.lock()
        defer { lock.unlock() }
        return currentId
    }

    private func setCurrent(_ id: String?) {
        lock.lock()
        currentId = id
        lock.unlock()
    }

    private func handle(_ op: Op) async {
        switch op {
        case let .player(sourceId, snapshot, appBundle, hasItem, wantsVolume):
            let s = sources[sourceId] ?? SourceState()
            s.snapshot = snapshot
            s.appBundle = appBundle
            s.hasItem = hasItem
            s.wantsVolume = wantsVolume
            seqCounter += 1
            s.seq = seqCounter
            sources[sourceId] = s
            await emitArbitrated()
        case let .queue(sourceId, queue):
            let s = sources[sourceId] ?? SourceState()
            s.queue = queue
            sources[sourceId] = s
            let cur = currentSource()
            if cur == nil || cur == sourceId {
                try? await gateway.player.queueChanged(queue)
            }
        case let .targets(sourceId, targets):
            let s = sources[sourceId] ?? SourceState()
            s.targets = targets
            sources[sourceId] = s
            let cur = currentSource()
            if cur == nil || cur == sourceId {
                try? await gateway.player.targetsChanged(targets)
            }
        case let .clear(sourceId):
            sources.removeValue(forKey: sourceId)
            if sourceId == currentSource() {
                setCurrent(nil)
                await emitArbitrated()
            }
        case .reconnect:
            heldScopes.removeAll()
            claimedBundle = nil
            await reemitCurrent()
        }
    }

    private func emitArbitrated() async {
        let prev = currentSource()
        let next = pickCurrent()
        setCurrent(next)
        guard let next, let s = sources[next] else {
            await releaseAll()
            return
        }
        if s.hasItem { await claim(s.appBundle, wantsVolume: s.wantsVolume) } else { await releaseAll() }
        if let snapshot = s.snapshot {
            try? await gateway.player.snapshot(snapshot)
        }
        if let prev, next != prev {
            try? await gateway.player.queueChanged(s.queue ?? QueueSnapshot(order: [], items: []))
            try? await gateway.player.targetsChanged(s.targets ?? PlaybackTargets(targets: []))
        }
    }

    private func reemitCurrent() async {
        var next = currentSource()
        if next == nil {
            next = pickCurrent()
            setCurrent(next)
        }
        guard let next, let s = sources[next] else { return }
        if s.hasItem { await claim(s.appBundle, wantsVolume: s.wantsVolume) }
        if let snapshot = s.snapshot { try? await gateway.player.snapshot(snapshot) }
        if let queue = s.queue { try? await gateway.player.queueChanged(queue) }
        if let targets = s.targets { try? await gateway.player.targetsChanged(targets) }
    }

    private func pickCurrent() -> String? {
        if sources.isEmpty { return nil }
        let playing = sources.filter { $0.value.hasItem && $0.value.snapshot?.playback.state == .playing }
        let pool = playing.isEmpty ? sources : playing
        return pool.max { a, b in a.value.seq < b.value.seq }?.key
    }

    private func claim(_ appBundle: String, wantsVolume: Bool) async {
        let bundleChanged = claimedBundle != appBundle
        var want = Self.nowPlayingScopes
        if wantsVolume { want.append(.volume) }
        for scope in want where !heldScopes.contains(scope) || bundleChanged {
            do {
                try await gateway.authority.claim(AuthorityClaim(scope: scope, appBundle: appBundle))
                heldScopes.insert(scope)
            } catch {
                continue
            }
        }
        if !wantsVolume, heldScopes.remove(.volume) != nil {
            try? await gateway.authority.release(AuthorityRelease(scope: .volume))
        }
        claimedBundle = appBundle
    }

    private func releaseAll() async {
        claimedBundle = nil
        guard !heldScopes.isEmpty else { return }
        let scopes = heldScopes
        heldScopes.removeAll()
        for scope in scopes {
            try? await gateway.authority.release(AuthorityRelease(scope: scope))
        }
    }

    private static let nowPlayingScopes: [CompanionAuthorityScope] = [
        .nowPlayingPlayback,
        .nowPlayingMetadata,
    ]
}

File: packages/companion/kotlin/companion/src/main/kotlin/com/bridgething/companion/NowPlayingHub.kt
package com.bridgething.companion

import com.bridgething.gateway.BridgethingGateway
import com.bridgething.gateway.authority
import com.bridgething.gateway.player
import com.bridgething.glue.NowPlayingSink
import com.bridgething.glue.NowPlayingTransport
import com.bridgething.schema.AuthorityClaim
import com.bridgething.schema.AuthorityRelease
import com.bridgething.schema.CompanionAuthorityScope
import com.bridgething.schema.PlaybackState
import com.bridgething.schema.PlaybackTargets
import com.bridgething.schema.PlayerState
import com.bridgething.schema.QueueSnapshot
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.launch

class NowPlayingHub(private val gateway: BridgethingGateway) : NowPlayingSink {
    private class SourceState {
        var snapshot: PlayerState? = null
        var appBundle: String = ""
        var hasItem: Boolean = false
        var wantsVolume: Boolean = false
        var queue: QueueSnapshot? = null
        var targets: PlaybackTargets? = null
        var seq: Long = 0
    }

    private sealed interface Op {
        data class Player(
            val sourceId: String,
            val snapshot: PlayerState,
            val appBundle: String,
            val hasItem: Boolean,
            val wantsVolume: Boolean,
        ) : Op
        data class Queue(val sourceId: String, val queue: QueueSnapshot) : Op
        data class Targets(val sourceId: String, val targets: PlaybackTargets) : Op
        data class Clear(val sourceId: String) : Op
        object Reconnect : Op
    }

    private val channel = Channel<Op>(Channel.UNLIMITED)
    private var consumer: Job? = null

    private val sources = HashMap<String, SourceState>()
    private var seqCounter: Long = 0
    @Volatile private var current: String? = null
    private val heldScopes = mutableSetOf<CompanionAuthorityScope>()
    private var claimedBundle: String? = null

    private val transports = java.util.concurrent.ConcurrentHashMap<String, NowPlayingTransport>()

    fun start(scope: CoroutineScope) {
        if (consumer != null) return
        consumer = scope.launch {
            for (op in channel) handle(op)
        }
    }

    fun stop() {
        consumer?.cancel()
        consumer = null
    }

    override fun submitPlayer(
        sourceId: String,
        snapshot: PlayerState,
        appBundle: String,
        hasItem: Boolean,
        wantsVolume: Boolean,
    ) {
        channel.trySend(Op.Player(sourceId, snapshot, appBundle, hasItem, wantsVolume))
    }

    override fun submitQueue(sourceId: String, queue: QueueSnapshot) {
        channel.trySend(Op.Queue(sourceId, queue))
    }

    override fun submitTargets(sourceId: String, targets: PlaybackTargets) {
        channel.trySend(Op.Targets(sourceId, targets))
    }

    override fun clearSource(sourceId: String) {
        channel.trySend(Op.Clear(sourceId))
    }

    fun onConnect() {
        channel.trySend(Op.Reconnect)
    }

    fun register(sourceId: String, transport: NowPlayingTransport) {
        transports[sourceId] = transport
    }

    fun unregister(sourceId: String) {
        transports.remove(sourceId)
    }

    fun currentTransport(): NowPlayingTransport? = current?.let { transports[it] }

    fun currentSource(): String? = current

    private suspend fun handle(op: Op) {
        when (op) {
            is Op.Player -> {
                val s = sources.getOrPut(op.sourceId) { SourceState() }
                s.snapshot = op.snapshot
                s.appBundle = op.appBundle
                s.hasItem = op.hasItem
                s.wantsVolume = op.wantsVolume
                s.seq = ++seqCounter
                emitArbitrated()
            }
            is Op.Queue -> {
                val s = sources.getOrPut(op.sourceId) { SourceState() }
                s.queue = op.queue
                if (current == null || current == op.sourceId) {
                    runCatching { gateway.player.queueChanged(op.queue) }
                }
            }
            is Op.Targets -> {
                val s = sources.getOrPut(op.sourceId) { SourceState() }
                s.targets = op.targets
                if (current == null || current == op.sourceId) {
                    runCatching { gateway.player.targetsChanged(op.targets) }
                }
            }
            is Op.Clear -> {
                sources.remove(op.sourceId)
                if (op.sourceId == current) {
                    current = null
                    emitArbitrated()
                }
            }
            Op.Reconnect -> {
                heldScopes.clear()
                claimedBundle = null
                reemitCurrent()
            }
        }
    }

    private suspend fun emitArbitrated() {
        val prev = current
        val next = pickCurrent()
        current = next
        if (next == null) {
            releaseAll()
            return
        }
        val s = sources.getValue(next)
        if (s.hasItem) claim(s.appBundle, s.wantsVolume) else releaseAll()
        s.snapshot?.let { runCatching { gateway.player.snapshot(it) } }
        if (prev != null && next != prev) {
            runCatching { gateway.player.queueChanged(s.queue ?: QueueSnapshot(order = emptyList(), items = emptyList())) }
            runCatching { gateway.player.targetsChanged(s.targets ?: PlaybackTargets(targets = emptyList())) }
        }
    }

    private suspend fun reemitCurrent() {
        val next = current ?: pickCurrent()?.also { current = it } ?: return
        val s = sources[next] ?: return
        if (s.hasItem) claim(s.appBundle, s.wantsVolume)
        s.snapshot?.let { runCatching { gateway.player.snapshot(it) } }
        s.queue?.let { runCatching { gateway.player.queueChanged(it) } }
        s.targets?.let { runCatching { gateway.player.targetsChanged(it) } }
    }

    private fun pickCurrent(): String? {
        if (sources.isEmpty()) return null
        val playing = sources.filter { it.value.hasItem && it.value.snapshot?.playback?.state == PlaybackState.Playing }
        val pool = if (playing.isNotEmpty()) playing else sources
        return pool.maxByOrNull { it.value.seq }?.key
    }

    private suspend fun claim(appBundle: String, wantsVolume: Boolean) {
        val bundleChanged = claimedBundle != appBundle
        val want = if (wantsVolume) NOW_PLAYING_SCOPES + CompanionAuthorityScope.Volume else NOW_PLAYING_SCOPES
        for (scope in want) {
            if (!heldScopes.contains(scope) || bundleChanged) {
                val sent = runCatching { gateway.authority.claim(AuthorityClaim(scope = scope, appBundle = appBundle)) }.isSuccess
                if (sent) heldScopes.add(scope)
            }
        }
        if (!wantsVolume && heldScopes.remove(CompanionAuthorityScope.Volume)) {
            runCatching { gateway.authority.release(AuthorityRelease(scope = CompanionAuthorityScope.Volume)) }
        }
        claimedBundle = appBundle
    }

    private suspend fun releaseAll() {
        claimedBundle = null
        if (heldScopes.isEmpty()) return
        val scopes = heldScopes.toList()
        heldScopes.clear()
        for (scope in scopes) {
            runCatching { gateway.authority.release(AuthorityRelease(scope = scope)) }
        }
    }

    private companion object {
        val NOW_PLAYING_SCOPES = listOf(
            CompanionAuthorityScope.NowPlayingPlayback,
            CompanionAuthorityScope.NowPlayingMetadata,
        )
    }
}
